Making some general fixs

shahzebsiddiqui opened the pull request:
- Update docstring where parameters are missing
- Reorder imports such that they are listed in alphabetical order
- Add some more logging during configuration check, and build process.
- Make use of is_file, is_dir when applicable for checking file path as
  pose to 'os' library
- In walk_tree if directory doesn't exist, we return with empty list
- Remove sub-directory "root"  in BUILDTEST_ROOT since it's not being
  used for anything, and refactor test
- Refactor test for walk_tree, add some print statement for
  troubleshooting, and test when invalid path is set. Also walk_tree is
  not going to raise an error for invalid directory it just returns an
  empty list

shahzebsiddiqui committed:
General fixs to code base that includes the following
- Update docstring where parameters are missing
- Reorder imports such that they are listed in alphabetical order
- Add some more logging during configuration check, and build process.
- Make use of is_file, is_dir when applicable for checking file path as
  pose to 'os' library
- In walk_tree if directory doesn't exist, we return with empty list
- Remove sub-directory "root"  in BUILDTEST_ROOT since it's not being
  used for anything, and refactor test
- Refactor test for walk_tree, add some print statement for
  troubleshooting, and test when invalid path is set. Also walk_tree is
  not going to raise an error for invalid directory it just returns an
  empty list

vsoch commented on tests/test_config.py:
```diff
@@ -39,5 +39,4 @@ def test_init_creation_of_buildtest_dir():
 
     shutil.rmtree(BUILDTEST_ROOT)
     init()
```
Since we know this is in a user's home, this test needs to run to be based in tmp_path. We can specify the BUILDTEST_ROOT to be there.

shahzebsiddiqui replied:
since ``init`` takes no argument the only way i can do this is set ``BUILDTEST_ROOT`` to ``tmp_path`` by updating variable in ``buildtest.defaults`` module so that init is using tmp_path. Do you think that what you want?

shahzebsiddiqui replied:
ok i was able to tweak the way we do ``init`` and creation process so that regression test can work with ``tmp_path`` see https://github.com/HPC-buildtest/buildtest-framework/pull/277/commits/e1f9598e3d8e49cb12ad2606dc6ba80561165b36 More importantly we can now pass any directory path where to store configuration file. Since 'site' is a sub-directory in BUILDTEST_ROOT, we could even alter where ``buildtest get``   will clone the repo. Generally user may not want stuff cloned in $HOME due to quota limitation.

vsoch commented on tests/utils/test_file.py:
```diff
@@ -57,12 +57,24 @@ def test_fail_create_dir():
-    walk_tree("/xyz", ".py")
+def test_walk_tree_no_files(tmp_path):
+    list_of_files = walk_tree(tmp_path, ".py")
+    print(f"Detected {len(list_of_files)} .py files found in directory: {here}")
```
Is this really testing walking a tree, if there will never be a tree in this location?

shahzebsiddiqui replied:
Ok there is a typo in the print statement, i fixed locally but yes it is doing a directory traversal in tmp_path expecting to find no .py files. This is different than the other test ``test_walk_tree_invalid_dir`` which is testing if invalid directory is specified. 
```
(buildtest-framework) ssi29@ag-mxg-hulk090> pytest -vrP tests/utils/test_file.py
============================================================================================= test session starts =============================================================================================
platform linux -- Python 3.7.4, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /u/users/ssi29/.local/share/virtualenvs/buildtest-framework-LDt8wyRf/bin/python3.7
cachedir: .pytest_cache
rootdir: /mxg-hpc/users/ssi29/buildtest-framework
collected 8 items

tests/utils/test_file.py::test_checking_directory PASSED                                                                                                                                                [ 12%]
tests/utils/test_file.py::test_checking_file XPASS                                                                                                                                                      [ 25%]
tests/utils/test_file.py::test_directory_expansion PASSED                                                                                                                                               [ 37%]
tests/utils/test_file.py::test_create_dir PASSED                                                                                                                                                        [ 50%]
tests/utils/test_file.py::test_fail_create_dir XFAIL                                                                                                                                                    [ 62%]
tests/utils/test_file.py::test_walk_tree PASSED                                                                                                                                                         [ 75%]
tests/utils/test_file.py::test_walk_tree_no_files PASSED                                                                                                                                                [ 87%]
tests/utils/test_file.py::test_walk_tree_invalid_dir PASSED                                                                                                                                             [100%]

=================================================================================================== PASSES ====================================================================================================
_______________________________________________________________________________________________ test_walk_tree ________________________________________________________________________________________________
-------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------
Detected 3 .py files found in directory: /mxg-hpc/users/ssi29/buildtest-framework/tests/utils
___________________________________________________________________________________________ test_walk_tree_no_files ___________________________________________________________________________________________
-------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------
Detected 0 .py files found in directory: /scratch/pytest-of-ssi29/pytest-1/test_walk_tree_no_files0
_________________________________________________________________________________________ test_walk_tree_invalid_dir __________________________________________________________________________________________
-------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------
Removing directory: /scratch/pytest-of-ssi29/pytest-1/test_walk_tree_invalid_dir0 first before doing directory traversal on invalid directory
Returned following files: [] with .py extension for path: /scratch/pytest-of-ssi29/pytest-1/test_walk_tree_invalid_dir0
=================================================================================== 6 passed, 1 xfailed, 1 xpassed in 0.93s ===================================================================================
```

shahzebsiddiqui committed:
add argument to method init, create_settings_file to pass settings_file
this allows for us to create arbitrary location where buildtest can
create the root directory and store the settings file. More importantly
we not bound to $HOME as done before.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -78,14 +79,15 @@ def load(self, buildspec):
             sys.exit("Buildspec File: %s does not exist." % self.buildspec)
 
-        elif os.path.isdir(self.buildspec):
+        elif is_dir(self.buildspec):
```
This change doesn't make sense - if you just checked with is_file this statement will never be reached. The original check was intended to first check if the path exists, and then parse if it's a directory or not (hitting the second statement). Now this is unreachable, unless there is some niche case of a symlink not being identified as a file and then hitting the next one.

shahzebsiddiqui replied:
Yea I see your point, the elif wont be reached i was testing this locally just for my sanity and hope we can adopt these methods as much as we can. 

is_file works with shell expansion and works by running ``os.path.isfile`` underneath
```
>>> is_file("$HOME/.buildtest/settings.yml")
True
>>> is_file("$HOME/.buildtest/")
False
```

If a directory is specified ``is_file`` will return ``False`` and the ``if not is_file`` will cause statment to be ``True`` and never reach the elif.

The thing is os.path.exists wouldn't work for some edge cases such as 

```
>>> os.path.exists("$HOME/.buildtest/settings.yml")
False
>>> is_file("$HOME/.buildtest/settings.yml")
True
>>> os.path.exists("$HOME/.buildtest/")
False
>>> is_dir("$HOME/.buildtest/")
True
```
We also account for tilde ``~`` if passed as input but os.path.exists won't catch this

```
>>> is_file("~/.buildtest/settings.yml")
True
>>> os.path.exists("~/.buildtest/settings.yml")
False
```
Symlinks work with os.path.exists so that's not an issue. It's only for discovery and exclusion we get the real path in-order for exclusion to work correctly.
The reason why i changed os.path.exists to ``is_file`` or ``is_dir`` depending on context of file is to account for tilde or shell expansion
```
>>> os.path.islink("/u/users/ssi29/gpfs")
True
>>> resolve_path("/u/users/ssi29/gpfs")
'/mxg-hpc/users/ssi29'
```

Note that ``is_file`` and ``is_dir`` are calling ``resolve_path`` before the check.

vsoch replied:
So under what conditions would the second elif trigger?

shahzebsiddiqui replied:
Yea i fixed this now in latest commit and let me demo this 

If input is a directory

```
>>> spec = BuildspecParser("$PWD/examples/benchmarks/")
Detected /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks is a directory, please provide a file path (not a directory path) to BuildspecParser.
```

If its invalid file

```
>>> spec = BuildspecParser("xyz!@#")
Can't process input: xyz!@#
```

If its valid file we proceed as normal

```
>>> spec = BuildspecParser("$PWD/examples/benchmarks/stream.yml")
>>> spec.recipe
{'version': '0.0.1', 'stream_uniprocess_c': {'type': 'script', 'description': 'STREAM Microbenchmark C test on uniprocessor', 'run': 'wget https://www.cs.virginia.edu/stream/FTP/Code/stream.c\ngcc -O2 stream.c -o stream\n./stream\n'}, 'stream_openmp_c': {'type': 'script', 'description': 'STREAM Microbenchmark C Test with OpenMP', 'run': 'wget https://www.cs.virginia.edu/stream/FTP/Code/stream.c\ngcc -fopenmp -D_OPENMP -O2 stream.c -o stream\n\nexport OMP_NUM_THREAD=2\n./stream\n\nexport OMP_NUM_THREAD=4\n./stream\n\nexport OMP_NUM_THREAD=8\n./stream\n\nexport OMP_NUM_THREAD=16\n./stream\n'}}

```

vsoch commented on buildtest/config.py:
```diff
@@ -3,42 +3,48 @@
-        os.mkdir(BUILDTEST_ROOT)
+    if not is_dir(dirname):
+        create_dir(dirname)
+        msg = f"Creating buildtest settings directory: {dirname}"
```
This seems redundant - if you want to print a message to the user and logger, use logger.info.

shahzebsiddiqui replied:
okay i can fix this

shahzebsiddiqui committed:
In BuildspecParser.load we use resolve_path to handle shell and user
expansion. We only exit if input is a directory or invalid file path.
resolve_path will return None if it's an invalid path.

shahzebsiddiqui committed:
applying black

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -82,15 +82,14 @@ def load(self, buildspec):
 
-        if not is_file(self.buildspec):
-            sys.exit("Buildspec File: %s does not exist." % self.buildspec)
+        self.buildspec = resolve_path(buildspec)
```
The check if buildspec is None needs to go _before_ this line - if buildspec here is None, it's going to spit out an error when you do expandvars in resolve_path.

vsoch replied:
E.g.,

```python
import os                                                                                                 
# This is the first line of resolve_path
path = os.path.expandvars(None)                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-4154076a32c3> in <module>
----> 1 path = os.path.expandvars(None)

~/anaconda3/lib/python3.7/posixpath.py in expandvars(path)
    286     """Expand shell variables of form $var and ${var}.  Unknown variables
    287     are left unchanged."""
--> 288     path = os.fspath(path)
    289     global _varprog, _varprogb
    290     if isinstance(path, bytes):

TypeError: expected str, bytes or os.PathLike object, not NoneType
```

shahzebsiddiqui replied:
Okay i made the change and tested this locally to ensure it hits all three conditionals. 

```
>>> spec = BuildspecParser(None)
Invalid input type for Buildspec, must be of type 'string'.
>>> spec = BuildspecParser("$PWD/invalid/file")
Can't process input: $PWD/invalid/file
>>> spec = BuildspecParser("$HOME/.buildtest")
Detected /u/users/ssi29/.buildtest is a directory, please provide a file path (not a directory path) to BuildspecParser.

```

shahzebsiddiqui replied:
@vsoch  i also added two more checks in ``test_base.py`` in commit https://github.com/shahzebsiddiqui/buildtest-framework/commit/b37a93a98f38f6e7728dca154a834bd5b7b7d93f to make sure we hit those conditionals in coverage report.

shahzebsiddiqui committed:
add condition if 'None' is passed to BuildspecParser, we terminate before calling 'resolve_path' because it's not expecting an input of type 'None'

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

shahzebsiddiqui committed:
adding test for 'None' and directory passed to BuildspecParser

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on tests/utils/test_file.py:
```diff
@@ -57,12 +57,24 @@ def test_fail_create_dir():
+def test_walk_tree_no_files(tmp_path):
+    list_of_files = walk_tree(tmp_path, ".py")
+    print(f"Detected {len(list_of_files)} .py files found in directory: {tmp_path}")
+    assert 0 == len(list_of_files)
```
This test could be improved to actually test that it's walking a directory (e.g., os.listdir()) would have worked here, there is no tree walking. But this is a nit, and otherwise the PR looks good.
